### PR TITLE
.NET Isolated middleware for orchestrator functions

### PR DESCRIPTION
**Pull request by cgillum**
This PR enables a cleaner coding experience for orchestrator functions in .NET Isolated.

**Before**:
```csharp
[Function(nameof(HelloCitiesUntyped))]
public static string HelloCitiesUntyped([OrchestrationTrigger] string requestState, FunctionContext functionContext) =>
    Microsoft.DurableTask.OrchestrationRunner.LoadAndRun<string, string>(requestState, async (context, _) =>
    {
        string result = "";
        result += await context.CallActivityAsync<string>(nameof(SayHelloUntyped), "Tokyo") + " ";
        result += await context.CallActivityAsync<string>(nameof(SayHelloUntyped), "London") + " ";
        result += await context.CallActivityAsync<string>(nameof(SayHelloUntyped), "Seattle");
        return result;
    },
    functionContext.InstanceServices);
```

**After**:
```csharp
[Function(nameof(HelloCitiesUntyped))]
public static async Task<string> HelloCitiesUntyped([OrchestrationTrigger] TaskOrchestrationContext context)
{
    string result = "";
    result += await context.CallActivityAsync<string>(nameof(SayHelloUntyped), "Tokyo") + " ";
    result += await context.CallActivityAsync<string>(nameof(SayHelloUntyped), "London") + " ";
    result += await context.CallActivityAsync<string>(nameof(SayHelloUntyped), "Seattle");
    return result;
}
```

It works by automatically injecting middleware into the .NET worker that does pretty much exactly what the "before" example was doing so that we could remove that boilerplate from user code.

Dependencies on both the .NET Isolated worker SDK and the new Durable Task .NET SDK (including its source generator) needed to be updated to support this.

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)
* [x] My changes **do not** require any extra work to be leveraged by OutOfProc SDKs
    * [ ] Otherwise: That work is being tracked here: #issue_or_pr_in_each_sdk
* [x] My changes **do not** change the version of the WebJobs.Extensions.DurableTask package
    * [ ] Otherwise: major or minor version updates are reflected in `/src/Worker.Extensions.DurableTask/AssemblyInfo.cs`

### CODE REVIEW

File: src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs [hunk 1]
```diff
@@ -98,7 +98,13 @@ public ValueTask<ConversionResult> ConvertAsync(ConverterContext converterContex
             try
             {
                 DurableTaskGrpcClient.Builder builder = DurableTaskGrpcClient.CreateBuilder();
-                builder.UseAddress(inputData.rpcBaseUrl);
+                if (!Uri.TryCreate(inputData.rpcBaseUrl, UriKind.Absolute, out Uri? baseUriResult))
+                {
+                    InvalidOperationException exception = new($"Failed to parse the {nameof(inputData.rpcBaseUrl)} field from the input binding data.");
+                    return new ValueTask<ConversionResult>(ConversionResult.Failed(exception));
+                }
```
Comment (cgillum) [inline src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs:105]: Not strictly required, but improving some validation based on some bugs I ran into during local development.

File: src/Worker.Extensions.DurableTask/DurableTaskExtensionStartup.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Azure.Functions.Worker.Core;
+using Microsoft.Azure.Functions.Worker.Extensions.DurableTask;
+using Microsoft.Extensions.Hosting;
+
+[assembly: WorkerExtensionStartup(typeof(DurableTaskExtensionStartup))]
```
Comment (cgillum) [inline src/Worker.Extensions.DurableTask/DurableTaskExtensionStartup.cs:8]: This line triggers the .NET worker to automatically run our `Configure` method below. This is how we inject our middleware without any customer action (besides referencing our extension).

File: src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        }
+
+        await orchestratorTask;
+
+        // This will throw if either the orchestrator performed an illegal await or if some middleware ahead of this
+        // one performed some illegal await.
+        orchestrationContext.ThrowIfIllegalAccess();
+    }
+
+    private sealed class FunctionsOrchestrationContext : TaskOrchestrationContext
```
Comment (cgillum) [inline src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs:96]: This is just a wrapper class for the orchestration context that checks to see if there are any illegal awaits or illegal async callbacks in the user code.

**Review by cgillum — COMMENTED** [review]
Adding some comments for additional context.

File: src/Worker.Extensions.DurableTask/Worker.Extensions.DurableTask.csproj [hunk 4]
```diff
@@ -39,12 +39,12 @@
   </PropertyGroup>
   
   <ItemGroup>
-    <PackageReference Include="Microsoft.Azure.Functions.Worker.Core" Version="1.5.0-preview1" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker.Core" Version="1.6.0-preview1" />
```
Comment (kshyju) [inline src/Worker.Extensions.DurableTask/Worker.Extensions.DurableTask.csproj:42]: FYI: preview3 is the [latest](https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker.Core/1.6.0-preview3)

Comment (cgillum) [reply src/Worker.Extensions.DurableTask/Worker.Extensions.DurableTask.csproj:42]: Oh, looks like I missed that! Any interesting improvements I might have missed in those two releases?

Comment (kshyju) [reply src/Worker.Extensions.DurableTask/Worker.Extensions.DurableTask.csproj:42]: No impact to the durable use cases. It was minor fixes in other areas.

File: src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs [hunk 5]
```diff
@@ -26,8 +26,15 @@ private DefaultDurableClientContext(DurableTaskClient client, DurableClientInput
...
-        ArgumentNullException.ThrowIfNull(inputData.requiredQueryStringParameters, nameof(inputData.requiredQueryStringParameters));
+        if (string.IsNullOrEmpty(inputData.taskHubName))
+        {
+            throw new ArgumentNullException(nameof(inputData.taskHubName));
+        }
+
+        if (string.IsNullOrEmpty(inputData.requiredQueryStringParameters))
+        {
+            throw new ArgumentNullException(nameof(inputData.requiredQueryStringParameters));
+        }
```
Comment (davidmrdavid) [inline src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs:37]: Does this change any behavior, or is this mostly just a style change?

File: src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+using System.Threading.Tasks;
+using Microsoft.Azure.Functions.Worker.Middleware;
+using Microsoft.DurableTask;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.Functions.Worker.Extensions.DurableTask;
+
+internal class DurableTaskFunctionsMiddleware : IFunctionsWorkerMiddleware
+{
+    // TODO: Need to add call to action, i.e., pointer to some documentation.
```
Comment (davidmrdavid) [inline src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs:16]: let's have them also file a bug report if they encounter this and cannot fix it :)

File: src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    public async Task Invoke(FunctionContext functionContext, FunctionExecutionDelegate next)
+    {
+        if (!IsOrchestrationTrigger(functionContext, out BindingMetadata? triggerMetadata))
+        {
+            await next(functionContext);
+            return;
+        }
+
+        InputBindingData<object> triggerInputData = await functionContext.BindInputAsync<object>(triggerMetadata);
+        if (triggerInputData?.Value is not string encodedOrchestratorState)
```
Comment (davidmrdavid) [inline src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs:30]: this C# feature is a little fancy. Is this checking that `triggerInputData.Value` does not contain a string, but if it does contain a string then that string gets assigned to `encoded OrchestratorState`, right?

File: src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        FunctionsOrchestrationContext orchestrationContext)
+    {
+        Task orchestratorTask = next(functionContext);
+        if (!orchestratorTask.IsCompleted && !orchestrationContext.IsAccessed)
+        {
+            // If the middleware returns before the orchestrator function's context object was accessed, then we
+            // know that either some middleware component went async or that the orchestrator function did some illegal
+            // await as its very first action.
+            throw new InvalidOperationException(IllegalAwaitErrorMessage);
+        }
```
Comment (davidmrdavid) [inline src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs:87]: Couldn't the orchestrator also complete, without accessing the context, if it is empty?

**Review by davidmrdavid — COMMENTED** [review]
Mostly LGTM, just left a few questions

File: src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs (hunk 5 above)
Comment (cgillum) [reply src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs:37]: Technically it's a behavior change. Before we'd only throw if it was null. Now we throw if it's null or empty.

File: src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs (hunk 7 above)
Comment (cgillum) [reply src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs:30]: It's effectively doing three things in one expression:
1. Declares a new local variable (called `encodedOrchestratorStatus` in this case)
2. Converts the object value into a `string`, if possible
3. Returns `true` if both the string conversion succeeded and the value is non-null.

It's a very convenient feature that helps remove boilerplate. :)

**Comment by cgillum** [comment]
@davidmrdavid thanks! Questions answered.

File: src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs (hunk 8 above)
Comment (davidmrdavid) [reply src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs:87]: Ah I see, I got that mixed up.

In that case, my only recommendation is to change the comment in this segment to read: "If the middleware returns *for an incomplete orchestrator* before the orchestrator function's context object was accessed, ...."

**Review by davidmrdavid — APPROVED** [review]
:shipit: 

But please see my comments recommendation on the illegal await check. It's in our thread above ^

**Comment by cgillum** [comment]
It pains me to make a minor comment-only update when we have a slow and semi-flakey CI, but okay :-P

**Pull request merged**